### ISSUE TITLE
(OneVsOneClassifier) Not able to convert sklearn model using pipeline to ONNX format for real time inferencing

### ISSUE THREAD

**Issue by pratikchhapolika** [onnx/sklearn-onnx#895]
It is a multi-class classification model with sklearn. 

I am using `OneVsOneClassifier` model to train and predict `150 intents`. Its a multi-class classification problem.

**Data:**

    text          intents
    
    text1         int1
    text2         int2

I convert these intents in labels using:

    le = LabelEncoder()
    y_train = le.fit_transform(y_train)
    y_test = le.fit_transform(y_test)

 **Expectation:**

Without changing the training pipeline or parameters, note the inference time. Currently, it's slow, ~1second for 1 inference. So to convert pipeline to `ONNX` format and then use for inferencing on 1 example.

**Code:**

    from sklearn.feature_extraction.text import TfidfVectorizer
    from sklearn.compose import ColumnTransformer
    from sklearn.multiclass import OneVsRestClassifier, OneVsOneClassifier
    from sklearn.pipeline import Pipeline
    from sklearn.svm import SVC,LinearSVC
    
    def create_pipe(clf):
        
        # Each pipeline uses the same column transformer.  
        column_trans = ColumnTransformer(
                [('Text', TfidfVectorizer(), 'text')
                 ],
                remainder='drop') 
        
        pipeline = Pipeline([('prep',column_trans),                     
                             ('clf', clf)])
         
        return pipeline
    
    def fit_and_print(pipeline):
        
        pipeline.fit(X_train, y_train)
        y_pred = pipeline.predict(X_test)
    
        print(metrics.classification_report(y_test, y_pred, 
                                            target_names=le.classes_, 
                                            digits=3))
    clf = OneVsOneClassifier(LinearSVC(random_state=42, class_weight='balanced'))
    pipeline = create_pipe(clf)
    %time fit_and_print(pipeline)
    
    # convert input to df
    
    def create_test_data(x):
        d = {'text' : x}
        df = pd.DataFrame(d, index=[0])
        return df
    
    revs=[]
    for idx in [948, 5717, 458]:
         cur = test.loc[idx, 'text']
         revs.append(cur)
    print(revs) 
    
    revs=sam['text'].values
    
    %%time
    for rev in revs:
        c_res = pipeline.predict(create_test_data(rev))
        print(rev, '=', labels[c_res[0]])

**ONNX conversion code**

    from skl2onnx import convert_sklearn
    from skl2onnx.common.data_types import FloatTensorType, StringTensorType
    
    initial_type = [('UTTERANCE', StringTensorType([None, 2]))]
    model_onnx = convert_sklearn(pipeline, initial_types=initial_type)

**Error**

    MissingShapeCalculator: Unable to find a shape calculator for type '<class 'sklearn.multiclass.OneVsOneClassifier'>'.
    It usually means the pipeline being converted contains a
    transformer or a predictor with no corresponding converter
    implemented in sklearn-onnx. If the converted is implemented
    in another library, you need to register
    the converted so that it can be used by sklearn-onnx (function
    update_registered_converter). If the model is not yet covered
    by sklearn-onnx, you may raise an issue to
    https://github.com/onnx/sklearn-onnx/issues
    to get the converter implemented or even contribute to the
    project. If the model is a custom model, a new converter must
    be implemented. Examples can be found in the gallery.

**How to resolve this ? Also how to do prediction after converting to ONNX format?**

**Comment by bmreiniger** [comment]
cross-referencing: https://stackoverflow.com/q/73153452/10495893

**Comment by pratikchhapolika** [comment]
@bmreiniger any help.

**Comment by xadupre** [comment]
The converter for OneVsOneClassifier has not been implemented yet. It should not be too complicated to do. In the meantime, OneVsRestClassifier has a converter.

**Comment by pratikchhapolika** [comment]
Hi @xadupre any help on this?

**Comment by xiaowuhu** [comment]
@pratikchhapolika I'm working on this converter now. Can you please share me some of your data (X_train, y_train, X_test, y_test, etc.) so I can do a fully testing?

**Comment by pratikchhapolika** [comment]
> @pratikchhapolika I'm working on this converter now. Can you please share me some of your data (X_train, y_train, X_test, y_test, etc.) so I can do a fully testing?

sure. I will share it by end of day today.

**Comment by pratikchhapolika** [comment]
> @pratikchhapolika I'm working on this converter now. Can you please share me some of your data (X_train, y_train, X_test, y_test, etc.) so I can do a fully testing?

@xiaowuhu Please find the data-set

```
**X_train**
Apply for smart connect

I had a payment of $10.74 come out on 9/6. However the order did not complete but I was still charged.

I ordered from abcd at 4:45 pm on Sunday, September 11 and paid through ABC. As soon as the transaction was complete through ABC and I was redirected to the website, the website said my credit card was declined. I am concerned I will not be receiving my product and yet I was charged for it. Can you cancel my transaction? The ABC receipt said payment was sent to bag.com. I assume they are affiliated? I did not receive any email from the website. I look forward to hearing from you. Thank you, Jones Friday

Wondering the issue with case id: PP-D-1194392777789

I need my hold money
```

```
**y_train**
SmartConnect
Unauth Transaction
Cancel Payment
Dispute Status
Dispute Status
```

```
**X_test**
Provide additional information Dispute Case ID: PP-D-46128254897. Need refund.
Damn! Not able to fit it in
Eed help disputing all the google play charges
How do I cancel a payment that was done as a guest
I cannot confirm now
```

```
**y_test**
Dispute Status
Dispute-Claim
No
Dispute Status
Cancel Payment
```

**Comment by pratikchhapolika** [comment]
> @pratikchhapolika I'm working on this converter now. Can you please share me some of your data (X_train, y_train, X_test, y_test, etc.) so I can do a fully testing?

@xiaowuhu 

I am using `OneVsOneClassifier` with `ColumnTransformer([('Text', TfidfVectorizer(), 'UTTERANCE')],remainder='drop') `

```
def create_pipe(clf):
    
    # Each pipeline uses the same column transformer.  
    column_trans = ColumnTransformer([('Text', TfidfVectorizer(), 'UTTERANCE')],remainder='drop') 
    
    pipeline = Pipeline([('prep',column_trans),                     
                         ('clf', clf)])
     
    return pipeline
```

```
clf = OneVsOneClassifier(LinearSVC(random_state=42, class_weight='balanced'))
pipeline = create_pipe(clf)
```
```
def fit_and_print(pipeline):
    
    pipeline.fit(X_train, y_train)
    y_pred = pipeline.predict(X_test)

    print(metrics.classification_report(y_test, y_pred, 
                                        target_names=le.classes_, 
                                        digits=3))
```

`%time fit_and_print(pipeline)`

**Comment by xiaowuhu** [comment]
got it. will update you when I am ready.

**Comment by pratikchhapolika** [comment]
> got it. will update you when I am ready.

sure @xiaowuhu Thank you!

**Comment by xiaowuhu** [comment]
@pratikchhapolika 

sorry to ask for the data format of X_train and others. Is it looks like pandas DataFrame:

```
    X_train = pd.DataFrame(
        {
            'UTTERANCE': [
                "Apply for smart connect",
                "I had a payment of $10.74 come out on 9/6. However the order did not complete but I was still charged.",
                "I ordered from abcd at 4:45 pm on Sunday, September 11 and paid through ABC. As soon as the transaction was complete through ABC and I was redirected to the website, the website said my credit card was declined. I am concerned I will not be receiving my product and yet I was charged for it. Can you cancel my transaction? The ABC receipt said payment was sent to bag.com. I assume they are affiliated? I did not receive any email from the website. I look forward to hearing from you. Thank you, Jones Friday",
                "Wondering the issue with case id: PP-D-1194392777789",
                "I need my hold money"]
        }
    )
```

or just a python list?

my goal is to make below code work:
```
    clf = OneVsOneClassifier(LinearSVC(random_state=42, class_weight='balanced'))
    column_trans = ColumnTransformer([('Text', TfidfVectorizer(), 'UTTERANCE')], remainder='drop')   
    pipeline = Pipeline([('prep', column_trans), ('clf', clf)])   
    pipeline.fit(X_train, y_train)
    y_pred = pipeline.predict(X_test)
    print(y_pred)
    return clf
```

**Comment by pratikchhapolika** [comment]
> @pratikchhapolika
> 
> sorry to ask for the data format of X_train and others. Is it looks like pandas DataFrame:
> 
> ```
>     X_train = pd.DataFrame(
>         {
>             'UTTERANCE': [
>                 "Apply for smart connect",
>                 "I had a payment of $10.74 come out on 9/6. However the order did not complete but I was still charged.",
>                 "I ordered from abcd at 4:45 pm on Sunday, September 11 and paid through ABC. As soon as the transaction was complete through ABC and I was redirected to the website, the website said my credit card was declined. I am concerned I will not be receiving my product and yet I was charged for it. Can you cancel my transaction? The ABC receipt said payment was sent to bag.com. I assume they are affiliated? I did not receive any email from the website. I look forward to hearing from you. Thank you, Jones Friday",
>                 "Wondering the issue with case id: PP-D-1194392777789",
>                 "I need my hold money"]
>         }
>     )
> ```
> 
> or just a python list?
> 
> my goal is to make below code work:
> 
> ```
>     clf = OneVsOneClassifier(LinearSVC(random_state=42, class_weight='balanced'))
>     column_trans = ColumnTransformer([('Text', TfidfVectorizer(), 'UTTERANCE')], remainder='drop')   
>     pipeline = Pipeline([('prep', column_trans), ('clf', clf)])   
>     pipeline.fit(X_train, y_train)
>     y_pred = pipeline.predict(X_test)
>     print(y_pred)
>     return clf
> ```

@xiaowuhu 

test= test[['UTTERANCE', 'INTENT']]
test.reset_index(inplace=True)


X_train = train[['UTTERANCE']]
X_test = test[['UTTERANCE']]

y_train = train['INTENT']
y_test= test['INTENT']

**Comment by xiaowuhu** [comment]
@pratikchhapolika We finished the OvO converter, please pip install from GitHub to have a try. We didn't release it to PiPy yet.

**Comment by pratikchhapolika** [comment]
> @pratikchhapolika We finished the OvO converter, please pip install from GitHub to have a try. We didn't release it to PiPy yet.

@xiaowuhu Could you please show, how can I use it in my use-case or example above? 
Could you please post the code here?

**Comment by pratikchhapolika** [comment]
@xiaowuhu Also please let us know how to install this particular `ovo` converter and use it.

**Comment by xiaowuhu** [comment]
> @xiaowuhu Also please let us know how to install this particular `ovo` converter and use it.

pip install git+https://github.com/onnx/sklearn-onnx

if you can uninstall the previous package coming from PyPI, it will be more clear to check the above installation.

**Comment by pratikchhapolika** [comment]
> > @pratikchhapolika We finished the OvO converter, please pip install from GitHub to have a try. We didn't release it to PiPy yet.
> 
> @xiaowuhu Could you please show, how can I use it in my use-case or example above? Could you please post the code here?

Please help with this also @xiaowuhu

**Comment by xiaowuhu** [comment]
@pratikchhapolika  sorry I cannot make your sample working even in sk-learn, so cannot test the converter.

we are using this test case: https://github.com/onnx/sklearn-onnx/blob/main/tests/test_sklearn_one_vs_one_classifier_converter.py

**Comment by pratikchhapolika** [comment]
> al_types=initial_type)

@xiaowuhu Thank you. Let me try on my notebook and check. Meanwhile is it fine to keep this open?

**Comment by pratikchhapolika** [comment]
> @pratikchhapolika
> 
> sorry to ask for the data format of X_train and others. Is it looks like pandas DataFrame:
> 
> ```

> ```
> 
> or just a python list?
> 
> my goal is to make below code work:
> 
> ```
>     clf = OneVsOneClassifier(LinearSVC(random_state=42, class_weight='balanced'))
>     column_trans = ColumnTransformer([('Text', TfidfVectorizer(), 'UTTERANCE')], remainder='drop')   
>     pipeline = Pipeline([('prep', column_trans), ('clf', clf)])   
>     pipeline.fit(X_train, y_train)
>     y_pred = pipeline.predict(X_test)
>     print(y_pred)
>     return clf
> ```

Hi @xiaowuhu , could you please delete the X_train data from comment above.  As it might violate some of my company's privacy.

**Comment by xiaowuhu** [comment]
@pratikchhapolika  done. please check.
